Question: how to use `method!` bind with a mutuable reference and `Ruby`

This question goes beyond my current understanding of traits, macros, and mutable references in Rust. I will describe the problem I am fixing with a code example. When implementing an I/O class for [OpenDAL](https://github.com/apache/opendal), I am having trouble understanding what Rust's compiler (rustc) is trying to tell me:

```rust
use std::io::Read;
use std::io::Write;

use magnus::class;
use magnus::method;
use magnus::prelude::*;
use magnus::Error;
use magnus::RModule;

use crate::*;

#[magnus::wrap(class = "OpenDAL::IO", free_immediately, size)]
pub struct OpenDALIO {
    reader: Option<ocore::StdReader>,
    writer: Option<ocore::StdWriter>,
    is_reader_open: bool,
    is_writer_open: bool,
}


pub fn format_io_error(err: std::io::Error) -> Error {
    Error::new(exception::standard_error(), err.to_string())
}

impl OpenDALIO {
    fn close_write(&mut self) -> Result<(), Error> {
        if self.is_writer_open {
            self.writer.as_mut().expect("writer is available").close().map_err(format_io_error)?;
            self.is_writer_open = false;
        }

        Ok(())
    }
}


pub fn include(gem_module: &RModule) -> Result<(), Error> {
    let class = gem_module.define_class("IO", class::object())?;
    // ...
    class.define_method("close_write", method!(OpenDALIO::close_write, 0))?;
}
```

rustc gives me the following error message:

```
error[E0599]: the method `call_handle_error` exists for fn item `fn(&mut OpenDALIO) -> Result<(), Error> {OpenDALIO::close_write}`, but its trait bounds were not satisfied
   --> src/opendal_io.rs:485:40
    |
485 |     class.define_method("close_write", method!(OpenDALIO::close_write, 0))?;
    |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ method cannot be called due to unsatisfied trait bounds
    |
    = note: the following trait bounds were not satisfied:
            `<for<'a> fn(&'a mut OpenDALIO) -> Result<(), magnus::Error> {OpenDALIO::close_write} as FnOnce<(&Ruby, _)>>::Output = _`
            which is required by `for<'a> fn(&'a mut OpenDALIO) -> Result<(), magnus::Error> {OpenDALIO::close_write}: RubyMethod0<_, _>`
            `for<'a> fn(&'a mut OpenDALIO) -> Result<(), magnus::Error> {OpenDALIO::close_write}: Fn<(&Ruby, _)>`
            which is required by `for<'a> fn(&'a mut OpenDALIO) -> Result<(), magnus::Error> {OpenDALIO::close_write}: RubyMethod0<_, _>`
            `<&for<'a> fn(&'a mut OpenDALIO) -> Result<(), magnus::Error> {OpenDALIO::close_write} as FnOnce<(&Ruby, _)>>::Output = _`
            which is required by `&for<'a> fn(&'a mut OpenDALIO) -> Result<(), magnus::Error> {OpenDALIO::close_write}: RubyMethod0<_, _>`
            `&for<'a> fn(&'a mut OpenDALIO) -> Result<(), magnus::Error> {OpenDALIO::close_write}: Fn<(&Ruby, _)>`
            which is required by `&for<'a> fn(&'a mut OpenDALIO) -> Result<(), magnus::Error> {OpenDALIO::close_write}: RubyMethod0<_, _>`
            `&mut for<'a> fn(&'a mut OpenDALIO) -> Result<(), magnus::Error> {OpenDALIO::close_write}: Fn<(_,)>`
            which is required by `&mut for<'a> fn(&'a mut OpenDALIO) -> Result<(), magnus::Error> {OpenDALIO::close_write}: Method0<_, _>`
            `<&mut for<'a> fn(&'a mut OpenDALIO) -> Result<(), magnus::Error> {OpenDALIO::close_write} as FnOnce<(&Ruby, _)>>::Output = _`
            which is required by `&mut for<'a> fn(&'a mut OpenDALIO) -> Result<(), magnus::Error> {OpenDALIO::close_write}: RubyMethod0<_, _>`
            `&mut for<'a> fn(&'a mut OpenDALIO) -> Result<(), magnus::Error> {OpenDALIO::close_write}: Fn<(&Ruby, _)>`
            which is required by `&mut for<'a> fn(&'a mut OpenDALIO) -> Result<(), magnus::Error> {OpenDALIO::close_write}: RubyMethod0<_, _>`
    = note: this error originates in the macro `method` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Regarding mutable references, does this mean magnus does not support mutable references for structs, or is something else causing this issue?

Regarding the `Ruby` struct, can I insert a `Ruby` struct into the method’s stack? I’m interested in obtaining the exception class.
Currently, I'm ussing `Error::new(magnus::exception::standard_error(), format!("OpenDAL doesn't support mode: {mode}"))` but the magnus documentation recommends getting exception classes from a `Ruby` struct.

--- a/README.md
+++ b/README.md
@@ -84,23 +84,25 @@ make sure the return type is `magnus::Value`.
 
 Magnus allows you to wrap Rust structs and enums as Ruby objects, enabling seamless interaction between Rust and Ruby. This functionality is ideal for exposing Rust logic to Ruby modules.
 
-Use one of approach to expose a Rust type to Ruby:
+Use one of the following approaches to expose a Rust type to Ruby:
 
-- A convenience macro [magnus::wrap](https://docs.rs/magnus/latest/magnus/attr.wrap.html)
-- More customized approach by implementing the [magnus::TypedData](https://docs.rs/magnus/latest/magnus/derive.TypedData.html) trait
-)
+* A convenience macro [`#[magnus::wrap]`].
+* More customised approach by implementing the [`magnus::TypedData`] trait.
+
+[`#[magnus::wrap]`]: https://docs.rs/magnus/latest/magnus/attr.wrap.html
+[`magnus::TypedData`]:  https://docs.rs/magnus/latest/magnus/derive.TypedData.html
 
 Then this Rust type can be:
 
-- Returned to Ruby as wrapped objects.
-- Passed back to Rust and automatically unwrapped to native Rust references.
+* Returned to Ruby as a wrapped object.
+* Passed back to Rust and automatically unwrapped to a native Rust reference.
 
 #### Basic Usage
 
 Here’s how you can wrap a simple Rust struct and expose its methods to Ruby:
 
 ```rust
-use magnus::{function, method, prelude::*, Ruby, Error};
+use magnus::{function, method, prelude::*, Error, Ruby};
 
 #[magnus::wrap(class = "Point")]
 struct Point {
@@ -139,7 +141,7 @@ fn init(ruby: &Ruby) -> Result<(), Error> {
 
 #### Handling Mutability
 
-Because Ruby GC manages memory to your Rust type Magnus, you can't use functions with mutable references. To allow mutable fields in wrapped Rust structs, you can use the newtype pattern with `RefCell`:
+Because Ruby's GC manages the memory where your Rust type is stored, Magnus can't bind functions with mutable references. To allow mutable fields in wrapped Rust structs, you can use the newtype pattern with `RefCell`:
 
 ```rust
 use std::cell::RefCell;
@@ -159,15 +161,17 @@ impl MutPoint {
 }
 ```
 
-Read the complete example [here](https://github.com/matsadler/magnus/blob/main/examples/mut_point.rs).
+See [`examples/mut_point.rs`] for the complete example.
+
+[`examples/mut_point.rs`]: https://github.com/matsadler/magnus/blob/main/examples/mut_point.rs
 
 #### Supporting Subclassing
 
-To enable Ruby subclassing for wrapped Rust types, the struct must:
+To enable Ruby subclassing for wrapped Rust types, the type must:
 
-- Implement the `Default` trait
-- Define an allocator
-- Define an initializer
+* Implement the `Default` trait.
+* Define an allocator.
+* Define an initialiser.
 
 ``` rust
 #[derive(Default)]
@@ -206,12 +210,13 @@ Use `magnus::Error` to propagate errors to Ruby from Rust:
 struct MutPoint(RefCell<Point>);
 
 impl MutPoint {
-    fn set_x(ruby: &Ruby, rb_self: &Self, i: i32) -> Result<(), Error> {
-        if i <= 0 {
-            return Err(Error::new(ruby.exception_arg_error(), "x must be positive"));
+    fn add_x(ruby: &Ruby, rb_self: &Self, val: isize) -> Result<isize, Error> {
+        if let Some(sum) = rb_self.0.borrow().x.checked_add(val) {
+            rb_self.0.borrow_mut().x = sum;
+            Ok(sum)
+        } else {
+            return Err(Error::new(ruby.exception_range_error(), "result out of range"));
         }
-        rb_self.0.borrow_mut().x = i as isize;
-        Ok(())
     }
 }
 ```

--- a/README.md
+++ b/README.md
@@ -82,16 +82,25 @@ make sure the return type is `magnus::Value`.
 
 ### Wrapping Rust Types in Ruby Objects
 
-Rust structs and enums can be wrapped in Ruby objects so they can be returned
-to Ruby.
+Magnus allows you to wrap Rust structs and enums as Ruby objects, enabling seamless interaction between Rust and Ruby. This functionality is ideal for exposing Rust logic to Ruby modules.
 
-Types can opt-in to this with the `magnus::wrap` macro (or by implementing
-`magnus::TypedData`). Whenever a compatible type is returned to Ruby it will be
-wrapped in the specified class, and whenever it is passed back to Rust it will
-be unwrapped to a reference.
+Use one of approach to expose a Rust type to Ruby:
+
+- A convenience macro [magnus::wrap](https://docs.rs/magnus/latest/magnus/attr.wrap.html)
+- More customized approach by implementing the [magnus::TypedData](https://docs.rs/magnus/latest/magnus/derive.TypedData.html) trait
+)
+
+Then this Rust type can be:
+
+- Returned to Ruby as wrapped objects.
+- Passed back to Rust and automatically unwrapped to native Rust references.
+
+#### Basic Usage
+
+Here’s how you can wrap a simple Rust struct and expose its methods to Ruby:
 
 ```rust
-use magnus::{function, method, prelude::*, Error, Ruby};
+use magnus::{function, method, prelude::*, Ruby, Error};
 
 #[magnus::wrap(class = "Point")]
 struct Point {
@@ -128,16 +137,20 @@ fn init(ruby: &Ruby) -> Result<(), Error> {
 }
 ```
 
-The newtype pattern and `RefCell` can be used if mutability is required:
+#### Handling Mutability
+
+Because Ruby GC manages memory to your Rust type Magnus, you can't use functions with mutable references. To allow mutable fields in wrapped Rust structs, you can use the newtype pattern with `RefCell`:
 
 ```rust
+use std::cell::RefCell;
+
 struct Point {
     x: isize,
     y: isize,
 }
 
 #[magnus::wrap(class = "Point")]
-struct MutPoint(std::cell::RefCell<Point>);
+struct MutPoint(RefCell<Point>);
 
 impl MutPoint {
     fn set_x(&self, i: isize) {
@@ -146,8 +159,15 @@ impl MutPoint {
 }
 ```
 
-To allow wrapped types to be subclassed they must implement `Default`, and
-define and alloc func and an initialize method:
+Read the complete example [here](https://github.com/matsadler/magnus/blob/main/examples/mut_point.rs).
+
+#### Supporting Subclassing
+
+To enable Ruby subclassing for wrapped Rust types, the struct must:
+
+- Implement the `Default` trait
+- Define an allocator
+- Define an initializer
 
 ``` rust
 #[derive(Default)]
@@ -174,6 +194,25 @@ fn init(ruby: &Ruby) -> Result<(), Error> {
     class.define_alloc_func::<MutPoint>();
     class.define_method("initialize", method!(MutPoint::initialize, 2))?;
     Ok(())
+}
+```
+
+#### Error Handling
+
+Use `magnus::Error` to propagate errors to Ruby from Rust:
+
+```rust
+#[magnus::wrap(class = "Point")]
+struct MutPoint(RefCell<Point>);
+
+impl MutPoint {
+    fn set_x(ruby: &Ruby, rb_self: &Self, i: i32) -> Result<(), Error> {
+        if i <= 0 {
+            return Err(Error::new(ruby.exception_arg_error(), "x must be positive"));
+        }
+        rb_self.0.borrow_mut().x = i as isize;
+        Ok(())
+    }
 }
 ```
 

--- a/examples/mut_point.rs
+++ b/examples/mut_point.rs
@@ -1,12 +1,34 @@
 use std::cell::RefCell;
 
-use magnus::{function, method, prelude::*, wrap};
+use magnus::{function, method, prelude::*, wrap, Error, Ruby};
 
 struct Point {
     x: isize,
     y: isize,
 }
-
+/// The `wrap` macro wraps a Rust type in Ruby, enabling seamless integration.
+/// Magnus uses two Ruby API functions to manage the struct:
+/// - `rb_data_typed_object_wrap`
+/// - `rb_check_typeddata`
+///
+/// # Mutability
+///
+/// Ruby's garbage collector (GC) manages memory for wrapped objects. This prevents using `&mut` references
+/// because Rust requires them to be unique, while Ruby GC may move objects unpredictably. To address this, you can use
+/// [`RefCell`](https://doc.rust-lang.org/std/cell/struct.RefCell.html) to enable interior mutability.
+///
+/// # Error Handling
+///
+/// Use [`magnus::Error`](https://docs.rs/magnus/latest/magnus/struct.Error.html) to propagate errors to Ruby.
+/// For example, you can raise a Ruby exception from Rust using Ruby's predefined exception classes.
+///
+/// The syntax for methods like `set_x` differs slightly from typical Rust struct methods because it uses the
+/// [`method!` macro](https://docs.rs/magnus/latest/magnus/macro.method.html):
+/// - The first parameter, `ruby`, gives access to Ruby's runtime.
+/// - The second parameter, `rb_self`, is the Ruby object being called.
+///
+/// @see [`DataTypeFunctions`](https://docs.rs/magnus/latest/magnus/derive.DataTypeFunctions.html)
+/// @see [`TypedData`](https://docs.rs/magnus/latest/magnus/derive.TypedData.html)
 #[wrap(class = "Point")]
 struct MutPoint(RefCell<Point>);
 
@@ -19,8 +41,15 @@ impl MutPoint {
         self.0.borrow().x
     }
 
-    fn set_x(&self, val: isize) {
-        self.0.borrow_mut().x = val;
+    /// Setter for the x coordinate, with validation
+    ///
+    /// Ensures `x` is positive and raises a Ruby `ArgumentError` if the condition is not met.
+    fn set_x(ruby: &Ruby, rb_self: &Self, i: i32) -> Result<(), Error> {
+        if i <= 0 {
+            return Err(Error::new(ruby.exception_arg_error(), "x must be positive"));
+        }
+        rb_self.0.borrow_mut().x = i as isize;
+        Ok(())
     }
 
     fn y(&self) -> isize {

--- a/examples/mut_point.rs
+++ b/examples/mut_point.rs
@@ -6,29 +6,37 @@ struct Point {
     x: isize,
     y: isize,
 }
-/// The `wrap` macro wraps a Rust type in Ruby, enabling seamless integration.
-/// Magnus uses two Ruby API functions to manage the struct:
-/// - `rb_data_typed_object_wrap`
-/// - `rb_check_typeddata`
-///
-/// # Mutability
-///
-/// Ruby's garbage collector (GC) manages memory for wrapped objects. This prevents using `&mut` references
-/// because Rust requires them to be unique, while Ruby GC may move objects unpredictably. To address this, you can use
-/// [`RefCell`](https://doc.rust-lang.org/std/cell/struct.RefCell.html) to enable interior mutability.
-///
-/// # Error Handling
-///
-/// Use [`magnus::Error`](https://docs.rs/magnus/latest/magnus/struct.Error.html) to propagate errors to Ruby.
-/// For example, you can raise a Ruby exception from Rust using Ruby's predefined exception classes.
-///
-/// The syntax for methods like `set_x` differs slightly from typical Rust struct methods because it uses the
-/// [`method!` macro](https://docs.rs/magnus/latest/magnus/macro.method.html):
-/// - The first parameter, `ruby`, gives access to Ruby's runtime.
-/// - The second parameter, `rb_self`, is the Ruby object being called.
-///
-/// @see [`DataTypeFunctions`](https://docs.rs/magnus/latest/magnus/derive.DataTypeFunctions.html)
-/// @see [`TypedData`](https://docs.rs/magnus/latest/magnus/derive.TypedData.html)
+
+// The `wrap` macro wraps a Rust type in a Ruby object, enabling seamless
+// integration.
+//
+// Magnus uses two Ruby API functions to manage the struct:
+// * `rb_data_typed_object_wrap`
+// * `rb_check_typeddata`
+//
+// # Mutability
+//
+// Ruby's garbage collector (GC) manages memory for wrapped objects. This
+// prevents using `&mut` references because Rust requires them to be unique,
+// while Ruby GC may move objects unpredictably. To address this, you can use
+// [`RefCell`](https://doc.rust-lang.org/std/cell/struct.RefCell.html) to
+// enable interior mutability.
+//
+// # Error Handling
+//
+// Use [`magnus::Error`](https://docs.rs/magnus/latest/magnus/struct.Error.html)
+// to propagate errors to Ruby.
+// For example, you can raise a Ruby exception from Rust using Ruby's
+// predefined exception classes.
+//
+// The syntax for methods like `add_x` differs slightly from typical Rust
+// struct methods because it uses the
+// [`method!` macro](https://docs.rs/magnus/latest/magnus/macro.method.html):
+// * The first parameter, `ruby`, gives access to Ruby's runtime.
+// * The second parameter, `rb_self`, is the Ruby object being called.
+//
+// See [`DataTypeFunctions`](https://docs.rs/magnus/latest/magnus/derive.DataTypeFunctions.html)
+// and [`TypedData`](https://docs.rs/magnus/latest/magnus/derive.TypedData.html)
 #[wrap(class = "Point")]
 struct MutPoint(RefCell<Point>);
 
@@ -41,15 +49,20 @@ impl MutPoint {
         self.0.borrow().x
     }
 
-    /// Setter for the x coordinate, with validation
-    ///
-    /// Ensures `x` is positive and raises a Ruby `ArgumentError` if the condition is not met.
-    fn set_x(ruby: &Ruby, rb_self: &Self, i: i32) -> Result<(), Error> {
-        if i <= 0 {
-            return Err(Error::new(ruby.exception_arg_error(), "x must be positive"));
+    fn set_x(&self, val: isize) {
+        self.0.borrow_mut().x = val;
+    }
+
+    fn add_x(ruby: &Ruby, rb_self: &Self, val: isize) -> Result<isize, Error> {
+        if let Some(sum) = rb_self.0.borrow().x.checked_add(val) {
+            rb_self.0.borrow_mut().x = sum;
+            Ok(sum)
+        } else {
+            return Err(Error::new(
+                ruby.exception_range_error(),
+                "result out of range",
+            ));
         }
-        rb_self.0.borrow_mut().x = i as isize;
-        Ok(())
     }
 
     fn y(&self) -> isize {
@@ -58,6 +71,18 @@ impl MutPoint {
 
     fn set_y(&self, val: isize) {
         self.0.borrow_mut().y = val;
+    }
+
+    fn add_y(ruby: &Ruby, rb_self: &Self, val: isize) -> Result<isize, Error> {
+        if let Some(sum) = rb_self.0.borrow().y.checked_add(val) {
+            rb_self.0.borrow_mut().y = sum;
+            Ok(sum)
+        } else {
+            return Err(Error::new(
+                ruby.exception_range_error(),
+                "result out of range",
+            ));
+        }
     }
 
     fn distance(&self, other: &MutPoint) -> f64 {
@@ -71,8 +96,10 @@ fn main() -> Result<(), String> {
         class.define_singleton_method("new", function!(MutPoint::new, 2))?;
         class.define_method("x", method!(MutPoint::x, 0))?;
         class.define_method("x=", method!(MutPoint::set_x, 1))?;
+        class.define_method("add_x", method!(MutPoint::add_x, 1))?;
         class.define_method("y", method!(MutPoint::y, 0))?;
         class.define_method("y=", method!(MutPoint::set_y, 1))?;
+        class.define_method("add_y", method!(MutPoint::add_y, 1))?;
         class.define_method("distance", method!(MutPoint::distance, 1))?;
 
         let d: f64 = ruby.eval(

--- a/magnus-macros/src/lib.rs
+++ b/magnus-macros/src/lib.rs
@@ -8,8 +8,8 @@
 //! }
 //!
 //! #[magnus::init]
-//! fn init(ruby: &Ruby) -> Result<(), magnus::Error> {
-//!     ruby.define_class("RbPoint", magnus::class::object())?;
+//! fn init(ruby: &magnus::Ruby) -> Result<(), magnus::Error> {
+//!     ruby.define_class("RbPoint", ruby.class_object())?;
 //!     Ok(())
 //! }
 //! ```
@@ -120,39 +120,40 @@ pub fn init(attrs: TokenStream, item: TokenStream) -> TokenStream {
 /// The `#[wrap]` macro supports several attributes to configure its behavior:
 ///
 /// * `class = "..."` (required):  
-///   Specifies the Ruby class associated with the Rust type. The class name can include
-///   module paths, such as `Foo::Bar::Baz`, which are used to define the hierarchy
-///   of the Ruby class.
+///   Specifies the Ruby class associated with the Rust type. Supports module
+///   paths, such as `Foo::Bar::Baz`, to locate nested classes.
 ///   
 /// * `name = "..."`:  
-///   Specifies a debug name for the type. This name must be unique and defaults to the
-///   class name if not explicitly provided.
+///   Specifies a debug name for the type. This name must be unique and
+///   defaults to the class name if not explicitly provided.
 ///
 /// * `free_immediately`:  
-///   Indicates that the Rust type should be dropped as soon as the Ruby object is garbage
-///   collected. This is only safe if the type's [`Drop`] implementation does not call Ruby,
-///   as calling Ruby during the garbage collection process is unsafe and can lead to
-///   undefined behavior.
+///   Indicates that the Rust type should be dropped as soon as the Ruby object
+///   is garbage collected. This is only safe if the type's [`Drop`]
+///   implementation does not call Ruby, as calling Ruby during the garbage
+///   collection process is unsafe and can lead to undefined behavior.
 ///
 /// * `size`:  
-///   Reports the [`std::mem::size_of_val`] of the type to Ruby, helping Ruby's garbage
-///   collector determine when to run.
+///   Reports the [`std::mem::size_of_val`] of the type to Ruby, helping Ruby's
+///   garbage collector determine when to run.
 ///
 /// * `unsafe_generics`:  
-///   Disables compile-time checks for types with generics, allowing their use with
-///   `#[wrap]`. This should only be used if you are confident that the derived
-///   implementation of [`TypedData`] is correct for your generic type.
+///   Disables compile-time checks for types with generics, allowing their use
+///   with `#[wrap]`. This should only be used if you are confident that the
+///   derived implementation of [`TypedData`] is correct for your generic type.
 ///
 /// # Variant Attributes
 ///
-/// When wrapping enums, the `#[magnus(...)]` attribute can also be applied to individual
-/// variants to define specific behavior for them:
+/// When wrapping enums, the `#[magnus(...)]` attribute can also be applied to
+/// individual variants to define specific behavior for them:
 ///
 /// * `class = "..."`:  
-///   Specifies the Ruby class associated with a particular variant. This is useful
-///   for defining subclasses for the variants.
+///   Specifies the Ruby class associated with a particular variant. This is
+///   useful for defining subclasses for the variants.
 ///
-/// ## Wrapping a Struct
+/// # Examples
+///
+/// Wrapping a struct:
 ///
 /// ```
 /// use magnus::{function, prelude::*, wrap};
@@ -178,13 +179,16 @@ pub fn init(attrs: TokenStream, item: TokenStream) -> TokenStream {
 /// }
 /// ```
 ///
-/// Read the complete example [here](https://github.com/matsadler/magnus/blob/main/examples/point.rs).
+/// See [`examples/point.rs`] for the complete example.
 ///
-/// ## Handling Mutability
+/// [`examples/point.rs`]: https://github.com/matsadler/magnus/blob/main/examples/point.rs
+///
+/// Handling mutability:
 ///
 /// ```
-/// use magnus::{wrap, prelude::*};
 /// use std::cell::RefCell;
+///
+/// use magnus::{prelude::*, wrap};
 ///
 /// struct Point {
 ///     x: isize,
@@ -201,13 +205,16 @@ pub fn init(attrs: TokenStream, item: TokenStream) -> TokenStream {
 /// }
 /// ```
 ///
-/// Read the complete example [here](https://github.com/matsadler/magnus/blob/main/examples/mut_point.rs).
+/// See [`examples/mut_point.rs`] for the complete example.
 ///
-/// ## Supporting Subclassing
+/// [`examples/mut_point.rs`]: https://github.com/matsadler/magnus/blob/main/examples/mut_point.rs
+///
+/// Supporting subclassing:
 ///
 /// ```
-/// use magnus::{function, method, prelude::*, wrap, Ruby, Error};
 /// use std::cell::RefCell;
+///
+/// use magnus::{function, method, prelude::*, wrap, Error, Ruby};
 ///
 /// #[derive(Default)]
 /// struct Point {
@@ -236,11 +243,12 @@ pub fn init(attrs: TokenStream, item: TokenStream) -> TokenStream {
 /// }
 /// ```
 ///
-/// ## Error Handling
+/// Error handling:
 ///
 /// ```
-/// use magnus::{wrap, prelude::*, Ruby, Error};
 /// use std::cell::RefCell;
+///
+/// use magnus::{prelude::*, wrap, Error, Ruby};
 ///
 /// struct Point {
 ///     x: isize,
@@ -251,21 +259,26 @@ pub fn init(attrs: TokenStream, item: TokenStream) -> TokenStream {
 /// struct MutPoint(RefCell<Point>);
 ///
 /// impl MutPoint {
-///     fn set_x(ruby: &Ruby, rb_self: &Self, i: i32) -> Result<(), Error> {
-///         if i <= 0 {
-///             return Err(Error::new(ruby.exception_arg_error(), "x must be positive"));
+///     fn add_x(ruby: &Ruby, rb_self: &Self, val: isize) -> Result<isize, Error> {
+///         if let Some(sum) = rb_self.0.borrow().x.checked_add(val) {
+///             rb_self.0.borrow_mut().x = sum;
+///             Ok(sum)
+///         } else {
+///             return Err(Error::new(
+///                 ruby.exception_range_error(),
+///                 "result out of range",
+///             ));
 ///         }
-///         rb_self.0.borrow_mut().x = i as isize;
-///         Ok(())
 ///     }
 /// }
 /// ```
 ///
-/// ## Wrapping an Enum with Subclasses
+/// Wrapping an enum with subclasses:
 ///
 /// ```
-/// use magnus::{class, define_class, function, method, prelude::*, wrap};
 /// use std::f64::consts::PI;
+///
+/// use magnus::{class, define_class, function, method, prelude::*, wrap};
 ///
 /// #[wrap(class = "Shape")]
 /// enum Shape {
@@ -282,6 +295,20 @@ pub fn init(attrs: TokenStream, item: TokenStream) -> TokenStream {
 ///             Shape::Rectangle { x, y } => x * y,
 ///         }
 ///     }
+/// }
+///
+/// #[magnus::init]
+/// fn init() -> Result<(), magnus::Error> {
+///     let shape = define_class("Shape", class::object())?;
+///     shape.define_method("area", method!(Shape::area, 0))?;
+///
+///     let circle = define_class("Circle", shape)?;
+///     circle.define_singleton_method("new", function!(|r| Shape::Circle { r }, 1))?;
+///
+///     let rectangle = define_class("Rectangle", shape)?;
+///     rectangle.define_singleton_method("new", function!(|x, y| Shape::Rectangle { x, y }, 2))?;
+///
+///     Ok(())
 /// }
 /// ```
 #[proc_macro_attribute]
@@ -300,27 +327,30 @@ pub fn derive_data_type_functions(input: TokenStream) -> TokenStream {
     typed_data::expand_derive_data_type_functions(parse_macro_input!(input)).into()
 }
 
-/// Derives `TypedData`, allowing a Rust type to be passed to Ruby and automatically
-/// wrapped as a Ruby object.
+/// Derives `TypedData`, allowing a Rust type to be passed to Ruby and
+/// automatically wrapped as a Ruby object.
 ///
 /// For simpler use cases, consider using [`macro@wrap`].
 ///
 /// # Attributes
 ///
-/// The `#[magnus(...)]` attribute supports the following values to configure its behavior:
+/// The `#[magnus(...)]` attribute supports the following values to configure
+/// its behavior:
 ///
 /// * `class = "..."` (required):  
-///   Specifies the Ruby class associated with the Rust type. Supports module paths,
-///   such as `Foo::Bar::Baz`.
+///   Specifies the Ruby class associated with the Rust type. Supports module
+///   paths, such as `Foo::Bar::Baz`.
 ///
 /// * `name = "..."`:  
-///   Specifies a debug name for the type. This name must be unique and defaults to the
-///   class name if not explicitly provided.
+///   Specifies a debug name for the type. This name must be unique and
+///   defaults to the class name if not explicitly provided.
 ///
 /// * `free_immediately`:  
-///   Indicates that the Rust type should be dropped as soon as the Ruby object is garbage
-///   collected. This is only safe if the type's [`Drop`] and `DataTypeFunctions::free`
-///   implementations do not call Ruby.
+///   Indicates that the Rust type should be dropped as soon as the Ruby object
+///   is garbage collected. This is only safe if the type's [`Drop`] and
+///   `DataTypeFunctions::free` implementations do not call Ruby.
+///   If safe this should be enabled as this performs better and is more
+///   memory efficient.
 ///
 /// * `mark`:  
 ///   Enables Ruby to call the `DataTypeFunctions::mark` function.
@@ -332,36 +362,51 @@ pub fn derive_data_type_functions(input: TokenStream) -> TokenStream {
 ///   Enables Ruby to call the `DataTypeFunctions::compact` function.
 ///
 /// * `wb_protected`:  
-///   Sets the `wb_protected` flag for write-barrier-protected objects.
+///   Enable the 'write barrier protected' flag.
+///
+///   Types that contain Ruby values by default do not participate in
+///   generational GC (they are scanned every GC). This flag asserts all
+///   operations that write Ruby values to this type are protected with
+///   write barriers (see `typed_data::Writebarrier::writebarrier`) so this
+///   type can participate in generational GC.
+///
+///   This is hard to get right, and it is recomended you do not use this
+///   flag.
 ///
 /// * `frozen_shareable`:  
-///   Sets the `frozen_shareable` flag for objects shareable across frozen Ruby objects.
+///   Sets the `frozen_shareable` flag for objects that can be safely accessed
+///   concurrently from multiple threads when frozen. This allows objects to
+///   shared between Ractors.
 ///
 /// * `unsafe_generics`:  
-///   Disables compile-time checks for types with generics, allowing their use with
-///   `#[magnus(...)]`. Use this only if you are confident the derived implementation
-///   is correct for your generic type.
+///   Disables compile-time checks for types with generics, allowing their use
+///   with `#[magnus(...)]`. Use this only if you are confident the derived
+///   implementation is correct for your generic type.
 ///
 /// # Field Attributes
 ///
-/// The `#[magnus(...)]` attribute can be set on struct fields with the following values:
+/// The `#[magnus(...)]` attribute can be set on struct fields with the
+/// following values:
 ///
 /// * `opaque_attr_reader`:  
-///   For a Ruby value wrapped in `Opaque`, creates an accessor method that returns
-///   the unwrapped Ruby value.
+///   For a Ruby value wrapped in `Opaque`, creates an accessor method that
+///   returns the unwrapped Ruby value.
 ///
 /// # Variant Attributes
 ///
-/// The `#[magnus(...)]` attribute can be set on enum variants with the following values:
+/// The `#[magnus(...)]` attribute can be set on enum variants with the
+/// following values:
 ///
 /// * `class = "..."`:  
-///   Specifies the Ruby class associated with the variant. Supports module paths,
-///   such as `Foo::Bar::Baz`.
+///   Specifies the Ruby class associated with the variant. Supports module
+///   paths, such as `Foo::Bar::Baz`.
 ///
-/// ## Wrapping a Struct
+/// # Examples
+///
+/// Wrapping a struct
 ///
 /// ```
-/// use magnus::{TypedData, DataTypeFunctions, function, prelude::*, Ruby};
+/// use magnus::{function, prelude::*, DataTypeFunctions, Ruby, TypedData};
 ///
 /// #[derive(DataTypeFunctions, TypedData)]
 /// #[magnus(class = "Point", size, free_immediately)]
@@ -388,10 +433,10 @@ pub fn derive_data_type_functions(input: TokenStream) -> TokenStream {
 /// }
 /// ```
 ///
-/// ## Wrapping an Enum with Subclasses
+/// Wrapping an enum with subclasses:
 ///
 /// ```
-/// use magnus::{class, prelude::*, wrap, Ruby, Error};
+/// use magnus::{class, prelude::*, wrap, Error, Ruby};
 ///
 /// #[wrap(class = "Shape")]
 /// enum Shape {
@@ -410,10 +455,11 @@ pub fn derive_data_type_functions(input: TokenStream) -> TokenStream {
 /// }
 /// ```
 ///
-/// ## Custom `DataTypeFunctions` Implementation
+/// Custom `DataTypeFunctions` implementation:
 ///
 /// ```
 /// use std::mem::size_of_val;
+///
 /// use magnus::{DataTypeFunctions, TypedData};
 ///
 /// #[derive(TypedData)]
@@ -430,12 +476,12 @@ pub fn derive_data_type_functions(input: TokenStream) -> TokenStream {
 /// }
 /// ```
 ///
-/// ## Struct Containing Ruby Values
+/// Struct containing Ruby values:
 ///
 /// ```
 /// use magnus::{
-///     class, function, gc, method, prelude::*, typed_data::Obj, value::Opaque,
-///     DataTypeFunctions, TypedData, Ruby
+///     class, function, gc, method, prelude::*, typed_data::Obj, value::Opaque, DataTypeFunctions,
+///     Ruby, TypedData,
 /// };
 ///
 /// #[derive(DataTypeFunctions, TypedData)]

--- a/magnus-macros/src/lib.rs
+++ b/magnus-macros/src/lib.rs
@@ -8,8 +8,8 @@
 //! }
 //!
 //! #[magnus::init]
-//! fn init() -> Result<(), magnus::Error> {
-//!     magnus::define_class("RbPoint", magnus::class::object())?;
+//! fn init(ruby: &Ruby) -> Result<(), magnus::Error> {
+//!     ruby.define_class("RbPoint", magnus::class::object())?;
 //!     Ok(())
 //! }
 //! ```
@@ -110,71 +110,164 @@ pub fn init(attrs: TokenStream, item: TokenStream) -> TokenStream {
     .into()
 }
 
-/// Allow a Rust type to be passed to Ruby, automatically wrapped as a Ruby
-/// object.
+/// Allows a Rust type to be passed to Ruby, where it is automatically wrapped
+/// as a Ruby object.
 ///
-/// For more control over the wrapped object, see [`TypedData`].
+/// For more control over the wrapping behavior, see [`TypedData`].
 ///
 /// # Attributes
 ///
-/// * `class = "..."` - required, sets the Ruby class to wrap the Rust type.
-///   Supports module paths, e.g. `Foo::Bar::Baz`.
-/// * `name = "..."` - debug name for the type, must be unique. Defaults to the
-///   class name.
-/// * `free_immediately` - Drop the Rust type as soon as the Ruby object has
-///   been garbage collected. This is only safe to set if the type's [`Drop`]
-///   implmentation does not call Ruby.
-/// * `size` - Report the [`std::mem::size_of_val`] of the type to Ruby, used
-///   to aid in deciding when to run the garbage collector.
-/// * `unsafe_generics` - The derived implementation of [`TypedData`] is not
-///   guaranteed to be correct for types with generics. If you are sure it is
-///   for your type this attribute can be used to override the compile time
-///   error usually generated for types with generics.
+/// The `#[wrap]` macro supports several attributes to configure its behavior:
+///
+/// * `class = "..."` (required):  
+///   Specifies the Ruby class associated with the Rust type. The class name can include
+///   module paths, such as `Foo::Bar::Baz`, which are used to define the hierarchy
+///   of the Ruby class.
+///   
+/// * `name = "..."`:  
+///   Specifies a debug name for the type. This name must be unique and defaults to the
+///   class name if not explicitly provided.
+///
+/// * `free_immediately`:  
+///   Indicates that the Rust type should be dropped as soon as the Ruby object is garbage
+///   collected. This is only safe if the type's [`Drop`] implementation does not call Ruby,
+///   as calling Ruby during the garbage collection process is unsafe and can lead to
+///   undefined behavior.
+///
+/// * `size`:  
+///   Reports the [`std::mem::size_of_val`] of the type to Ruby, helping Ruby's garbage
+///   collector determine when to run.
+///
+/// * `unsafe_generics`:  
+///   Disables compile-time checks for types with generics, allowing their use with
+///   `#[wrap]`. This should only be used if you are confident that the derived
+///   implementation of [`TypedData`] is correct for your generic type.
 ///
 /// # Variant Attributes
 ///
-/// The `#[magnus(...)]` attribute can be set on enum variants with the
-/// following values:
+/// When wrapping enums, the `#[magnus(...)]` attribute can also be applied to individual
+/// variants to define specific behavior for them:
 ///
-/// * `class = "..."` - sets the Ruby class to wrap the variant. Supports
-///   module paths, e.g. `Foo::Bar::Baz`.
+/// * `class = "..."`:  
+///   Specifies the Ruby class associated with a particular variant. This is useful
+///   for defining subclasses for the variants.
 ///
-/// # Examples
+/// ## Wrapping a Struct
 ///
 /// ```
-/// #[magnus::wrap(class = "RbPoint", free_immediately, size)]
+/// use magnus::{function, prelude::*, wrap};
+///
+/// #[wrap(class = "Point", free_immediately, size)]
 /// struct Point {
 ///     x: isize,
 ///     y: isize,
 /// }
 ///
-/// // the `Point` struct is automatically wrapped in a Ruby `RbPoint` object
-/// // when returned to Ruby.
 /// fn point(x: isize, y: isize) -> Point {
 ///     Point { x, y }
 /// }
 ///
-/// // Ruby `RbPoint` objects are automatically unwrapped to references to the
-/// // `Point` structs they are wrapping when this function is called from Ruby.
 /// fn distance(a: &Point, b: &Point) -> f64 {
 ///     (((b.x - a.x).pow(2) + (b.y - a.y).pow(2)) as f64).sqrt()
 /// }
 ///
 /// #[magnus::init]
 /// fn init() {
-///     magnus::define_global_function("point", magnus::function!(point, 2));
-///     magnus::define_global_function("distance", magnus::function!(distance, 2));
+///     magnus::define_global_function("point", function!(point, 2));
+///     magnus::define_global_function("distance", function!(distance, 2));
 /// }
 /// ```
 ///
-/// With subclasses for enum variants:
+/// Read the complete example [here](https://github.com/matsadler/magnus/blob/main/examples/point.rs).
+///
+/// ## Handling Mutability
 ///
 /// ```
+/// use magnus::{wrap, prelude::*};
+/// use std::cell::RefCell;
+///
+/// struct Point {
+///     x: isize,
+///     y: isize,
+/// }
+///
+/// #[wrap(class = "Point")]
+/// struct MutPoint(RefCell<Point>);
+///
+/// impl MutPoint {
+///     fn set_x(&self, i: isize) {
+///         self.0.borrow_mut().x = i;
+///     }
+/// }
+/// ```
+///
+/// Read the complete example [here](https://github.com/matsadler/magnus/blob/main/examples/mut_point.rs).
+///
+/// ## Supporting Subclassing
+///
+/// ```
+/// use magnus::{function, method, prelude::*, wrap, Ruby, Error};
+/// use std::cell::RefCell;
+///
+/// #[derive(Default)]
+/// struct Point {
+///     x: isize,
+///     y: isize,
+/// }
+///
+/// #[derive(Default)]
+/// #[wrap(class = "Point")]
+/// struct MutPoint(RefCell<Point>);
+///
+/// impl MutPoint {
+///     fn initialize(&self, x: isize, y: isize) {
+///         let mut this = self.0.borrow_mut();
+///         this.x = x;
+///         this.y = y;
+///     }
+/// }
+///
+/// #[magnus::init]
+/// fn init(ruby: &Ruby) -> Result<(), Error> {
+///     let class = ruby.define_class("Point", ruby.class_object()).unwrap();
+///     class.define_alloc_func::<MutPoint>();
+///     class.define_method("initialize", method!(MutPoint::initialize, 2))?;
+///     Ok(())
+/// }
+/// ```
+///
+/// ## Error Handling
+///
+/// ```
+/// use magnus::{wrap, prelude::*, Ruby, Error};
+/// use std::cell::RefCell;
+///
+/// struct Point {
+///     x: isize,
+///     y: isize,
+/// }
+///
+/// #[wrap(class = "Point")]
+/// struct MutPoint(RefCell<Point>);
+///
+/// impl MutPoint {
+///     fn set_x(ruby: &Ruby, rb_self: &Self, i: i32) -> Result<(), Error> {
+///         if i <= 0 {
+///             return Err(Error::new(ruby.exception_arg_error(), "x must be positive"));
+///         }
+///         rb_self.0.borrow_mut().x = i as isize;
+///         Ok(())
+///     }
+/// }
+/// ```
+///
+/// ## Wrapping an Enum with Subclasses
+///
+/// ```
+/// use magnus::{class, define_class, function, method, prelude::*, wrap};
 /// use std::f64::consts::PI;
 ///
-/// use magnus::{class, define_class, function, method, prelude::*};
-///
-/// #[magnus::wrap(class = "Shape")]
+/// #[wrap(class = "Shape")]
 /// enum Shape {
 ///     #[magnus(class = "Circle")]
 ///     Circle { r: f64 },
@@ -189,20 +282,6 @@ pub fn init(attrs: TokenStream, item: TokenStream) -> TokenStream {
 ///             Shape::Rectangle { x, y } => x * y,
 ///         }
 ///     }
-/// }
-///
-/// #[magnus::init]
-/// fn init() -> Result<(), magnus::Error> {
-///     let shape = define_class("Shape", class::object())?;
-///     shape.define_method("area", method!(Shape::area, 0))?;
-///
-///     let circle = define_class("Circle", shape)?;
-///     circle.define_singleton_method("new", function!(|r| Shape::Circle { r }, 1))?;
-///
-///     let rectangle = define_class("Rectangle", shape)?;
-///     rectangle.define_singleton_method("new", function!(|x, y| Shape::Rectangle { x, y }, 2))?;
-///
-///     Ok(())
 /// }
 /// ```
 #[proc_macro_attribute]
@@ -221,85 +300,100 @@ pub fn derive_data_type_functions(input: TokenStream) -> TokenStream {
     typed_data::expand_derive_data_type_functions(parse_macro_input!(input)).into()
 }
 
-/// Derives `TypedData`, allowing the type to be passed to Ruby automatically
+/// Derives `TypedData`, allowing a Rust type to be passed to Ruby and automatically
 /// wrapped as a Ruby object.
 ///
-/// For simple cases, see [`macro@wrap`].
+/// For simpler use cases, consider using [`macro@wrap`].
 ///
 /// # Attributes
 ///
-/// The `#[magnus(...)]` attribute can be set with the following values:
+/// The `#[magnus(...)]` attribute supports the following values to configure its behavior:
 ///
-/// * `class = "..."` - required, sets the Ruby class to wrap the Rust type.
-///   Supports module paths, e.g. `Foo::Bar::Baz`.
-/// * `name = "..."` - debug name for the type, must be unique. Defaults to the
-///   class name.
-/// * `free_immediately` - Drop the Rust type as soon as the Ruby object has
-///   been garbage collected. This is only safe to set if the type's [`Drop`]
-///   and `DataTypeFunctions::free` implementations do not call Ruby.
-/// * `mark` - Enable Ruby calling the `DataTypeFunctions::mark` function.
-/// * `size` - Enable Ruby calling the `DataTypeFunctions::size` function.
-/// * `compact` - Enable Ruby calling the `DataTypeFunctions::compact` function.
-/// * `wb_protected` - Enable the `wb_protected` flag.
-/// * `frozen_shareable` - Enable the `frozen_shareable` flag.
-/// * `unsafe_generics` - The derived implementation of [`TypedData`] is not
-///   guaranteed to be correct for types with generics. If you are sure it is
-///   for your type this attribute can be used to override the compile time
-///   error usually generated for types with generics.
+/// * `class = "..."` (required):  
+///   Specifies the Ruby class associated with the Rust type. Supports module paths,
+///   such as `Foo::Bar::Baz`.
+///
+/// * `name = "..."`:  
+///   Specifies a debug name for the type. This name must be unique and defaults to the
+///   class name if not explicitly provided.
+///
+/// * `free_immediately`:  
+///   Indicates that the Rust type should be dropped as soon as the Ruby object is garbage
+///   collected. This is only safe if the type's [`Drop`] and `DataTypeFunctions::free`
+///   implementations do not call Ruby.
+///
+/// * `mark`:  
+///   Enables Ruby to call the `DataTypeFunctions::mark` function.
+///
+/// * `size`:  
+///   Enables Ruby to call the `DataTypeFunctions::size` function.
+///
+/// * `compact`:  
+///   Enables Ruby to call the `DataTypeFunctions::compact` function.
+///
+/// * `wb_protected`:  
+///   Sets the `wb_protected` flag for write-barrier-protected objects.
+///
+/// * `frozen_shareable`:  
+///   Sets the `frozen_shareable` flag for objects shareable across frozen Ruby objects.
+///
+/// * `unsafe_generics`:  
+///   Disables compile-time checks for types with generics, allowing their use with
+///   `#[magnus(...)]`. Use this only if you are confident the derived implementation
+///   is correct for your generic type.
 ///
 /// # Field Attributes
 ///
-/// The `#[magnus(...)]` attribute can be set on struct fields with the
-/// following values:
+/// The `#[magnus(...)]` attribute can be set on struct fields with the following values:
 ///
-/// * `opaque_attr_reader` - For a Ruby value wrapped in `Opaque`, creates a
-///   accessor method that returns the unwrapped Ruby value.
+/// * `opaque_attr_reader`:  
+///   For a Ruby value wrapped in `Opaque`, creates an accessor method that returns
+///   the unwrapped Ruby value.
 ///
 /// # Variant Attributes
 ///
-/// The `#[magnus(...)]` attribute can be set on enum variants with the
-/// following values:
+/// The `#[magnus(...)]` attribute can be set on enum variants with the following values:
 ///
-/// * `class = "..."` - sets the Ruby class to wrap the variant. Supports
-///   module paths, e.g. `Foo::Bar::Baz`.
+/// * `class = "..."`:  
+///   Specifies the Ruby class associated with the variant. Supports module paths,
+///   such as `Foo::Bar::Baz`.
 ///
-/// # Examples
+/// ## Wrapping a Struct
 ///
 /// ```
-/// use magnus::{DataTypeFunctions, TypedData};
+/// use magnus::{TypedData, DataTypeFunctions, function, prelude::*, Ruby};
 ///
 /// #[derive(DataTypeFunctions, TypedData)]
-/// #[magnus(class = "RbPoint", size, free_immediately)]
+/// #[magnus(class = "Point", size, free_immediately)]
 /// struct Point {
 ///     x: isize,
 ///     y: isize,
 /// }
 ///
-/// // the `Point` struct is automatically wrapped in a Ruby `RbPoint` object
-/// // when returned to Ruby.
+/// // wraps `Point` struct in a Ruby `Point` object when returned to Ruby.
 /// fn point(x: isize, y: isize) -> Point {
 ///     Point { x, y }
 /// }
 ///
-/// // Ruby `RbPoint` objects are automatically unwrapped to references to the
-/// // `Point` structs they are wrapping when this function is called from Ruby.
+/// // automatically unwraps `Point` objects to references to `Point`` structs
+/// // when invoking the function.
 /// fn distance(a: &Point, b: &Point) -> f64 {
 ///     (((b.x - a.x).pow(2) + (b.y - a.y).pow(2)) as f64).sqrt()
 /// }
 ///
 /// #[magnus::init]
-/// fn init() {
-///     magnus::define_global_function("point", magnus::function!(point, 2));
-///     magnus::define_global_function("distance", magnus::function!(distance, 2));
+/// fn init(ruby: &Ruby) {
+///     ruby.define_global_function("point", function!(point, 2));
+///     ruby.define_global_function("distance", function!(distance, 2));
 /// }
 /// ```
 ///
-/// With subclasses for enum variants:
+/// ## Wrapping an Enum with Subclasses
 ///
 /// ```
-/// use magnus::{class, define_class};
+/// use magnus::{class, prelude::*, wrap, Ruby, Error};
 ///
-/// #[magnus::wrap(class = "Shape")]
+/// #[wrap(class = "Shape")]
 /// enum Shape {
 ///     #[magnus(class = "Circle")]
 ///     Circle { r: f64 },
@@ -308,19 +402,18 @@ pub fn derive_data_type_functions(input: TokenStream) -> TokenStream {
 /// }
 ///
 /// #[magnus::init]
-/// fn init() -> Result<(), magnus::Error> {
-///     let shape = define_class("Shape", class::object())?;
-///     define_class("Circle", shape)?;
-///     define_class("Rectangle", shape)?;
+/// fn init(ruby: &Ruby) -> Result<(), Error> {
+///     let shape = ruby.define_class("Shape", ruby.class_object())?;
+///     ruby.define_class("Circle", shape)?;
+///     ruby.define_class("Rectangle", shape)?;
 ///     Ok(())
 /// }
 /// ```
 ///
-/// Defining a custom `DataType` function:
+/// ## Custom `DataTypeFunctions` Implementation
 ///
 /// ```
 /// use std::mem::size_of_val;
-///
 /// use magnus::{DataTypeFunctions, TypedData};
 ///
 /// #[derive(TypedData)]
@@ -337,20 +430,21 @@ pub fn derive_data_type_functions(input: TokenStream) -> TokenStream {
 /// }
 /// ```
 ///
-/// A struct containing Ruby values.
+/// ## Struct Containing Ruby Values
 ///
 /// ```
 /// use magnus::{
-///     class, define_class, function, gc, method, prelude::*, typed_data::Obj, value::Opaque,
-///     DataTypeFunctions, TypedData,
+///     class, function, gc, method, prelude::*, typed_data::Obj, value::Opaque,
+///     DataTypeFunctions, TypedData, Ruby
 /// };
 ///
-/// # #[magnus::wrap(class = "Point", free_immediately, size)]
-/// # struct Point {
-/// #     x: isize,
-/// #     y: isize,
-/// # }
-/// #
+/// #[derive(DataTypeFunctions, TypedData)]
+/// #[magnus(class = "Point", size, free_immediately)]
+/// struct Point {
+///     x: isize,
+///     y: isize,
+/// }
+///
 /// #[derive(TypedData)]
 /// #[magnus(class = "Line", free_immediately, mark)]
 /// struct Line {
@@ -384,8 +478,8 @@ pub fn derive_data_type_functions(input: TokenStream) -> TokenStream {
 /// }
 ///
 /// #[magnus::init]
-/// fn init() -> Result<(), magnus::Error> {
-///     let line = define_class("Line", class::object())?;
+/// fn init(ruby: &Ruby) -> Result<(), magnus::Error> {
+///     let line = ruby.define_class("Line", ruby.class_object())?;
 ///     line.define_singleton_method("new", function!(Line::new, 2))?;
 ///     line.define_method("length", method!(Line::length, 0))?;
 ///     Ok(())

--- a/src/typed_data.rs
+++ b/src/typed_data.rs
@@ -303,7 +303,7 @@ where
     /// Types that contain Ruby values by default do not participate in
     /// generational GC (they are scanned every GC). This flag asserts all
     /// operations that write Ruby values to this type are protected with
-    /// write barriers (see [`typed_data::Writebarrier::writebarrier`]) so this
+    /// write barriers (see [`Writebarrier::writebarrier`]) so this
     /// type can participate in generational GC.
     ///
     /// This is hard to get right, and it is recomended you do not use this

--- a/tests/time.rs
+++ b/tests/time.rs
@@ -1,6 +1,6 @@
-use magnus::rb_assert;
-use magnus::{Error, Ruby};
 use std::time::SystemTime;
+
+use magnus::{rb_assert, Error, Ruby};
 
 #[test]
 fn test_all() {

